File: utils/src/main/java/io/github/aemogie/timble/utils/events/Event.java
package io.github.aemogie.timble.utils.events;

import java.util.Map.Entry;
import java.util.stream.Stream;

public class Event {
	
	protected boolean fire() {
		return getApplicableListeners().allMatch(listener -> listener.getValue().fire(this) || !listener.getValue().important);
	}
	
	protected Stream<Entry<Class<Event>, Listener<Event>>> getApplicableListeners() {
		return EventBus.getApplicableListeners(getClass());
	}
	
	public abstract static class Listener<T extends Event> {
		public final boolean important;
		
		protected Listener(boolean important) {this.important = important;}
		
		protected abstract boolean onFire(T event);
		
		@SuppressWarnings({"unchecked"})
		protected boolean fire(Event event) {
			try {
				return onFire((T) event);
			} catch (ClassCastException e) {
				e.printStackTrace();
				return false;
			}
		}
	}
}

File: utils/src/main/kotlin/io/github/aemogie/timble/utils/events/Event.kt
package io.github.aemogie.timble.utils.events

open class Event {
	open fun initFire(): Boolean {
		return true
	}
	
	open fun fire(listeners: ArrayList<Listener<Event?>>): Boolean {
		return listeners.stream().allMatch { listener: Listener<Event?> -> listener.fire(this) || !listener.important }
	}
	
	abstract class Listener<T : Event?> protected constructor(val important: Boolean) {
		abstract fun fire(event: T): Boolean
	}
}